ocaml/xapi/xapi_pool.ml

```
@@ -111,29 +111,33 @@ let pre_join_checks ~__context ~rpc ~session_id ~force =
-      error "Pool.join failed because of patches mismatch";
-      error "Remote has %s" (string_of_patches pool_patches);
-      error "Local has %s" (string_of_patches host_patches);
-      raise (Api_errors.Server_error(Api_errors.pool_hosts_not_homogeneous,
-                                     [(Printf.sprintf "Patches applied differ: Remote has %s -- Local has %s"
-                                         (string_of_patches pool_patches) (string_of_patches host_patches))]))
+  let assert_api_version_matches () =
+    let master = get_master rpc session_id in
+    let candidate_slave = Helpers.get_localhost ~__context in
+    let master_major = Client.Host.get_API_version_major ~rpc ~session_id ~self:master in
+    let master_minor = Client.Host.get_API_version_minor ~rpc ~session_id ~self:master in
+    let slave_major = Db.Host.get_API_version_major ~__context ~self:candidate_slave in
+    let slave_minor = Db.Host.get_API_version_minor ~__context ~self:candidate_slave in
+    if master_major <> slave_major || master_minor <> slave_minor then
+    begin
+      error "Master API version is %Ld.%Ld while candidate slave is %Ld.%Ld"
+        master_major master_minor slave_major slave_minor;
+      raise (Api_errors.Server_error(Api_errors.pool_joining_host_must_have_same_api_version, []))
+    end
+  in
+
+  let assert_db_schema_matches () =
+    let master = get_master rpc session_id in
+    let candidate_slave = Helpers.get_localhost ~__context in
+    let master_sw_version = Client.Host.get_software_version ~rpc ~session_id ~self:master in
+    let slave_sw_version = Db.Host.get_software_version ~__context ~self:candidate_slave in
+    let master_db_schema = try List.assoc Xapi_globs._db_schema master_sw_version with _ -> "" in
+    let slave_db_schema = try List.assoc Xapi_globs._db_schema slave_sw_version with _ -> "" in
+    if master_db_schema <> "" && slave_db_schema <> "" && master_db_schema <> slave_db_schema then
```

I think that it is better to raise an error.